packages/plugin-ext/src/main/browser/menus/menus-contribution-handler.ts

```diff
@@ -123,14 +123,12 @@ export class MenusContributionPointHandler {
                     if (!action.command) {
                         continue;
                     }
-                    toDispose.push(this.registerTitleAction(location, { ...action, when: undefined }, {
+                    toDispose.push(this.registerTitleAction(location, action, {
                         execute: widget => widget instanceof PluginViewWidget && this.commands.executeCommand(action.command!),
                         isEnabled: widget => widget instanceof PluginViewWidget &&
-                            this.viewContextKeys.with({ view: widget.options.viewId }, () =>
-                                this.commands.isEnabled(action.command!) && this.viewContextKeys.match(action.when)),
+                            this.commands.isEnabled(action.command!) && this.viewContextKeys.match(action.when),
```

Comment:
These lines seem *almost* equivalent. What was going wrong before with the `.with` check?

Reply:
After removing the `when: undefined`. On contextKey change (eg: view change) now toolBarItem.onDidChange will be fired (This will not happen before). 

Due to this there will be lot of updateToolbar being called. isVisible will be called for every command on focus.
This will in-turn update the context key and fire toolbar update.

Also the final UI was not correct, the toolbar items were not visible on focus, rather was available only after entering into the view.

Reply:
I have experimented with modifying these lines to look like this:
```
                    toDispose.push(this.registerTitleAction(location, action, {
                        execute: widget => widget instanceof PluginViewWidget && this.commands.executeCommand(action.command!),
                        isEnabled: widget => widget instanceof PluginViewWidget &&
                            this.viewContextKeys.with({ view: widget.options.viewId }, () =>
                                this.commands.isEnabled(action.command!) && this.viewContextKeys.match(action.when)),
                        isVisible: widget => widget instanceof PluginViewWidget &&
                            this.viewContextKeys.with({ view: widget.options.viewId }, () =>
                                this.commands.isVisible(action.command!) && this.viewContextKeys.match(action.when))
                    }));
```
and that seems to work - the scenarios described for testing have the correct behavior - and there are no shadows. It seems that what this achieves is that the widget used to check visibility is always the correct widget, and the context-key change `onmouseenter` happens in the background to ensure that the toolbars update appropriately. Does that seem correct to you?

Reply:
Yes looks correct. My only concern is on the recursive events fired due to setting of context keys on isVisible or isEnabled. 

Consider the following scenario like,
Commands available - npm reload (view: npm), maven flat (view: maven)
This isVisible or isEnabled is going to be called for both these commands. 
Previously we didn't fire onDidChange event so there was no issues in rendering. 
But now since we have onDidChange first a toolbar update will be fired for setting npm view and then for maven view. This will cause multiple re-renders and will increase with number of commands available. 

From debugging, this is what i found. Do correct me if my understanding is wrong.

Reply:
Hm... it seems to me that we *shouldn't* fire an event or set a context key because an event just because a particular check (`isVisible` or `isEnabled`) is run. Thing should run like this
```
Context changes in the application
  causes: context key changes
    causes: context key service fires event
      causes: TabbarToolbarRegistry recalculates visibility given new context
        causes: TabbarToolbarRegistry fires didChange
          causes: Toolbars update
```
And the flow should stop there. The tabbar toolbar registry updating shouldn't cause a context key change event, and the toolbars updating shouldn't cause a toolbar registry change event. If we have a situation where these events are looping into one another, we probably need to do a deeper adjustment of the logic. Have you been able to track down what causes the looping behavior you describe?